cmd/launcher/launcher.go

```diff
@@ -39,8 +45,174 @@ var (
+		)
+		flSocket = flagset.String(
+			"socket",
+			env.String("SOCKET", ""),
+			"The path to the socket",
+		)
+	)
+	flagset.Usage = commandUsage(flagset, "launcher query")
+	if err := flagset.Parse(args); err != nil {
+		return err
+	}
+
+	queries := queryFile{}
+
+	if _, err := os.Stat(*flQueries); err == nil {
+		data, err := ioutil.ReadFile(*flQueries)
+		if err != nil {
+			return errors.Wrap(err, "reading supplied queries file")
+		}
+		if err := json.Unmarshal(data, &queries); err != nil {
+			return errors.Wrap(err, "unmarshaling queries file json")
+		}
+	}
+
+	if *flQueries == "" {
+		stdinQueries, err := ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return errors.Wrap(err, "reading stdin")
+		}
+		if err := json.Unmarshal(stdinQueries, &queries); err != nil {
+			return errors.Wrap(err, "unmarshaling stdin queries json")
+		}
+	}
+
+	if *flSocket == "" {
+		return errors.New("--socket must be defined")
+	}
+
+	client, err := osquerygo.NewClient(*flSocket, 5*time.Second)
+	if err != nil {
+		return errors.Wrap(err, "opening osquery client connection on "+*flSocket)
+	}
+	defer client.Close()
+
+	results := struct {
+		Results map[string]interface{} `json:"results"`
+	}{
+		Results: map[string]interface{}{},
+	}
+
+	for name, query := range queries.Queries {
+		resp, err := client.Query(query)
+		if err != nil {
+			return errors.Wrap(err, "running query")
+		}
+
+		if resp.Status.Code != int32(0) {
+			fmt.Println("Error running query:", resp.Status.Message)
+		}
+
+		results.Results[name] = resp.Response
+	}
+
+	enc := json.NewEncoder(os.Stdout)
+	enc.SetIndent("", "    ")
+	if err := enc.Encode(results); err != nil {
+		panic(err)
+	}
+
+	return nil
+}
+
+func runSocket(args []string) error {
+	flagset := flag.NewFlagSet("launcher socket", flag.ExitOnError)
+	var (
+		flPath = flagset.String(
+			"path",
+			env.String("SOCKET_PATH", filepath.Join(os.TempDir(), "osquery.sock")),
+			"The path to the socket",
+		)
+	)
+	flagset.Usage = commandUsage(flagset, "launcher socket")
+	if err := flagset.Parse(args); err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(filepath.Dir(*flPath)); os.IsNotExist(err) {
+		if err := os.Mkdir(filepath.Dir(*flPath), fs.DirMode); err != nil {
+			return errors.Wrap(err, "creating socket path base directory")
+		}
+	}
+
+	runner, err := osquery.LaunchInstance(
+		osquery.WithExtensionSocketPath(*flPath),
+	)
+	if err != nil {
+		return errors.Wrap(err, "creating osquery instance")
+	}
+
+	fmt.Println(*flPath)
+
+	// Wait forever
+	sig := make(chan os.Signal)
+	signal.Notify(sig, os.Interrupt, os.Kill, syscall.Signal(15))
+	<-sig
+
+	// allow for graceful termination.
+	runner.Shutdown()
+
+	return nil
+}
+
 func main() {
 	logger := kolidelog.NewLogger(os.Stderr)
+
+	// if the launcher is being ran with a positional argument, handle that
+	// argument. If a known positional argument is not supplied, fall-back to
+	// running an osquery instance.
+	if len(os.Args) > 1 {
```

Comment:
Main should check  `if len(os.Args) < 2 {` and print usage.

Reply:
It's not a failure case if `len(os.Args)` is less than two, you might just be running the launcher with all your options configured via environment variables